### PR TITLE
Adding Insert and Update State Info States Api Endpoint

### PR DESCRIPTION
**Pull request by ccaprio97**
## What’s changing?

Adding Insert and Update State Info States Api Endpoint

## Why?

NAC-1646

## This PR has:

- [X] **JIRA ticket(s)** (_Must be listed in PR request and branch name should follow convention NAC-123-description_)<description>

- [ ] **Security Related Changes** (_Security Tab questions have been answered on JIRA ticket(s)_)

- [X] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [ ] **Documented API changes** (_required on public classes and methods, and elsewhere as appropriate_)
    - [ ] External API documentation - OpenApi docs and Generated Docs (e.g. https://github.com/18F/piipan/tree/dev/docs/openapi) 
    - [ ] Internal API documentation (e.g. https://github.com/18F/piipan/tree/dev/metrics/docs/openapi)   

- [ ] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [x] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)
    - [ ] Have IaC scripts run successfully against an existing deployed environment?
    - [ ] Have IaC scripts run successfully against a newly deployed environment?
    - [ ] Do the IaC changes require deleting existing resources?

## Anything else?

_Add other details that are helpful for review here_

## Code Review Checklist:

- [ ] Automated checks all pass? (code builds successfully, tests pass, test coverage meets/exceeds threshold, etc.) 
- [ ] Have all requirements and Acceptance Criteria been met?
- **If there are code changes**
    - [ ] Is there no unnecessary complexity? Code is readable, maintainable, etc.
    - [ ] Do the changes follow SOLID principles?
    - [ ] Are the changes robust? Concurrency considered? Performance? Security (e.g. no sql injections)?
    - [ ] Are newly added dependencies needed and appropriate? Are dependency files (e.g. packages.json & packages.lock.json) updated?
    - [ ] For any new projects/solutions, have they been added to code coverage files?
    - [ ] Public classes & methods are documented    
- **If there are API changes**
    - [ ] Does API make sense? Useful and concise?
    - [ ] Are existing API interfaces/contracts intact? i.e. no breaking changes
    - [ ] Are API tests included?
    - [ ] Are API responses appropriate and documented? (200s, 201, 404s, etc.)
    - [ ] Are internals kept private and not leaked into the API?
    - [ ] External API documentation - OpenApi docs and Generated Docs 
    - [ ] Internal API documentation    
- **If there are IAC changes**
    - [ ] IAC Scripts are readable & maintainable
    - [ ] IAC Scripts are documented
    - [ ] IAC scripts pass required checks
    - [ ] IAC scripts were run successfully by author
- **General Logging Questions**
    - [ ] Do new capabilities (or changed capabilities) include logging?
    - [ ] Is PII not logged?
- **General Error Handling Questions**
    - [ ] Has exception handling been added for new capabilities (or changed capabilities)?
    - [ ] Are errors/exceptions properly communicated to end-users? Were there requirements for user facing error messages?
    - [ ] Are errors/exceptions logged?
    - [ ] Is PII kept out of error/exception logging?
- **If there are Developer Documentation Changes**
    - [ ] General Project documentation (under piipan\docs. e.g. iac.md, architecture.md, etc.)
    - [ ] Github documentation additions/changes (Readmes)
- **General automated test questions**
    - [ ] End-to-End Tests provided (if code implements new user story)
    - [ ] Are tests sufficient?
    - [ ] Corner cases considered?
    - [ ] Are performance tests needed?

### CODE REVIEW

File: states/src/Piipan.States.Func.Api/packages.lock.json [hunk 1]
```diff
@@ -462,10 +473,10 @@
       },
       "Microsoft.Extensions.FileProviders.Abstractions": {
         "type": "Transitive",
-        "resolved": "2.2.0",
-        "contentHash": "EcnaSsPTqx2MGnHrmWOD0ugbuuqVT8iICqSqPzi45V5/MA1LjUNb0kwgcxBGqizV1R+WeBK7/Gw25Jzkyk9bIw==",
+        "resolved": "6.0.0",
```
Comment (bgfoshay) [inline states/src/Piipan.States.Func.Api/packages.lock.json:476]: Question: this is a major version bump. Just checking it was intentional as part of this PR?

File: states/src/Piipan.States.Func.Api/packages.lock.json [hunk 2]
```diff
@@ -462,10 +473,10 @@
       },
       "Microsoft.Extensions.FileProviders.Abstractions": {
         "type": "Transitive",
-        "resolved": "2.2.0",
-        "contentHash": "EcnaSsPTqx2MGnHrmWOD0ugbuuqVT8iICqSqPzi45V5/MA1LjUNb0kwgcxBGqizV1R+WeBK7/Gw25Jzkyk9bIw==",
+        "resolved": "6.0.0",
+        "contentHash": "0pd4/fho0gC12rQswaGQxbU34jOS1TPS8lZPpkFCH68ppQjHNHYle9iRuHeev1LhrJ94YPvzcRd8UmIuFk23Qw==",
         "dependencies": {
-          "Microsoft.Extensions.Primitives": "2.2.0"
+          "Microsoft.Extensions.Primitives": "6.0.0"
```
Comment (bgfoshay) [inline states/src/Piipan.States.Func.Api/packages.lock.json:479]: Question: this is a major version bump. Just checking it was intentional as part of this PR?

File: states/src/Piipan.States.Func.Api/packages.lock.json [hunk 3]
```diff
@@ -496,13 +507,12 @@
       },
       "Microsoft.Extensions.Hosting.Abstractions": {
         "type": "Transitive",
-        "resolved": "2.2.0",
-        "contentHash": "+k4AEn68HOJat5gj1TWa6X28WlirNQO9sPIIeQbia+91n03esEtMSSoekSTpMjUzjqtJWQN3McVx0GvSPFHF/Q==",
+        "resolved": "6.0.0",
```
Comment (bgfoshay) [inline states/src/Piipan.States.Func.Api/packages.lock.json:510]: Question: this is a major version bump. Just checking it was intentional as part of this PR?

File: states/src/Piipan.States.Func.Api/UpsertState.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+using Piipan.Shared.Parsers;
+using Piipan.Shared.Cryptography;
+using Piipan.Shared.Http;
+using Piipan.States.Api.Models;
+using Piipan.States.Core.DataAccessObjects;
+
+namespace Piipan.States.Func.Api
+{
+    /// <summary>
+    /// Azure Function implementing orchestrator matching API.
```
Comment (bgfoshay) [inline states/src/Piipan.States.Func.Api/UpsertState.cs:19]: Nitpick: This comment seems like maybe it isn't right?

File: states/src/Piipan.States.Func.Api/UpsertState.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        public UpsertState(IStateInfoDao stateInfoDao, IStreamParser<StateInfoRequest> requestParser, ICryptographyClient cryptographyClient)
+        {
+            Dapper.DefaultTypeMap.MatchNamesWithUnderscores = true;
+            _stateInfoDao = stateInfoDao;
+            _requestParser = requestParser;
+            _cryptographyClient = cryptographyClient;
+        }
+
+        /// <summary>
+        /// API endpoint for conducting matches across all participating states
```
Comment (bgfoshay) [inline states/src/Piipan.States.Func.Api/UpsertState.cs:36]: Nitpick: This comment also doesn't seem correct.

File: states/src/Piipan.States.Func.Api/UpsertState.cs (hunk 4 above)
Comment (ccaprio97) [reply states/src/Piipan.States.Func.Api/UpsertState.cs:19]: thats from a copy and paste section. will update both

File: states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Piipan.Shared.Parsers;
+using Piipan.States.Api.Models;
+
+namespace Piipan.States.Core.Parsers
+{
+    /// <summary>
+    /// Parser for deserializing AddEventRequest objects from a Stream
```
Comment (bgfoshay) [inline states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs:13]: Nitpick: more incorrect comments

File: states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        /// </remarks>
+        /// <param name="stream">A Stream from which a serialized AddEventRequest can be read</param>
+        /// <returns>A validated instance of AddEventRequest</returns>
+        public async Task<(StateInfoRequest ParsedRequest, JObject RawRequest)> Parse(Stream stream)
+        {
+            try
+            {
+                StateInfoRequest request = null;
+
+                var reader = new StreamReader(stream);
```
Comment (bgfoshay) [inline states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs:41]: Suggestion: use the reader in a using block so that it's disposed of properly

File: states/src/Piipan.States/Piipan.States.Core/DataAccessObjects/StateInfoDao.cs [hunk 8]
```diff
@@ -76,11 +80,56 @@ public async Task<IEnumerable<IState>> GetStates()
...
+                id = insertingState.Id,
+                state = insertingState.State,
+                state_abbreviation = insertingState.StateAbbreviation,
+                email = insertingState.Email,
+                phone = insertingState.Phone,
+                region = insertingState.Region,
+                email_cc = insertingState.EmailCc,
+            };
+
+            using (var connection = await _dbConnectionFactory.Build())
```
Comment (dborgen2) [inline states/src/Piipan.States/Piipan.States.Core/DataAccessObjects/StateInfoDao.cs:129]: Can you update this to use the DabaseManager pattern? We shouldn't be using dbConnectionFactory moving forward, but we haven't deleted it yet.

Comment (ccaprio97) [reply states/src/Piipan.States/Piipan.States.Core/DataAccessObjects/StateInfoDao.cs:129]: is that something I could add a .5 ticket to this sprint and do? if it doesnt break anything else then i can do it but if it turns into some bigger problem again then this ticket is really behind and has a lot of files

Comment (dborgen2) [reply states/src/Piipan.States/Piipan.States.Core/DataAccessObjects/StateInfoDao.cs:129]: It's literally changing
`using (var connection = await _dbConnectionFactory.Build())`
to
`return await _databaseManager.PerformQuery(async (connection) =>`

I think it should be fine to do as part of this

File: states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs (hunk 7 above)
Comment (ccaprio97) [reply states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs:41]: should this be updated in all the parsers because this is copied from addEventRequest Parser

File: states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+using Piipan.States.Func.Api.IntegrationTests;
+using System.Data.Common;
+using Xunit;
+
+namespace Piipan.States.Func.Api.IntegrationTests
+{
+    [Collection("StateApiTests")]
+    public class UpsertStateInfoApiIntegrationTests : DbFixture
+    {
+        private const string base64EncodedKey = "kW6QuilIQwasK7Maa0tUniCdO+ACHDSx8+NYhwCo7jQ=";
```
Comment (dborgen2) [inline states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs:29]: Question: Just wanted to verify that this key isn't anything that should be kept secret.

File: states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+                    new AzureServiceTokenProvider(),
+                    Environment.GetEnvironmentVariable(CollaborationDatabaseConnectionString)
+                );
+            });
+            Environment.SetEnvironmentVariable("ColumnEncryptionKey", base64EncodedKey);
+            services.RegisterKeyVaultClientServices();
+
+            var provider = services.BuildServiceProvider();
+
+            var key = "kW6QuilIQwasK7Maa0tUniCdO+ACHDSx8+NYhwCo7jQ=";
```
Comment (dborgen2) [inline states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs:68]: Nitpick: Isn't this the same key as base64EncodedKey? Do we need it twice?

File: states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+            {
+                return new BasicPgConnectionFactory<StateInfoDb>(
+                    Environment.GetEnvironmentVariable(Startup.CollaborationDatabaseConnectionString));
+            });
+            Environment.SetEnvironmentVariable("ColumnEncryptionKey", base64EncodedKey);
+            services.RegisterKeyVaultClientServices();
+
+            var provider = services.BuildServiceProvider();
+
+            var key = "kW6QuilIQwasK7Maa0tUniCdO+ACHDSx8+NYhwCo7jQ=";
```
Comment (dborgen2) [inline states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs:105]: Nitpick: Same here, I think this is the same key as base64EncodedKey. We can probably just use that.

File: states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs (hunk 7 above)
Comment (bgfoshay) [reply states/src/Piipan.States/Piipan.States.Core/Parsers/StateInfoRequestParser.cs:41]: No, I'll create a ticket for it to revisit. Don't worry about it now.

File: states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs (hunk 9 above)
Comment (ccaprio97) [reply states/tests/Piipan.States.Func.Api.IntegrationTests/UpsertStateInfoApiIntegrationTests.cs:29]: i dont think so. its kept in the other tests that have cryptography

File: states/src/Piipan.States/Piipan.States.Core/DataAccessObjects/StateInfoDao.cs (hunk 8 above)
Comment (ccaprio97) [reply states/src/Piipan.States/Piipan.States.Core/DataAccessObjects/StateInfoDao.cs:129]: gotcha. wasnt sure how big it was.

**Review by bgfoshay — APPROVED** [review]
👍 Nice work getting this one done. I know there was a lot of scope creep that factored into this work, but I think we're better positioned to handle state info changes now as well as having building blocks available for creating an administration page.
